service/create/domains.go

```diff
@@ -0,0 +1,111 @@
+package create
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/giantswarm/aws-operator/resources"
+	awsresources "github.com/giantswarm/aws-operator/resources/aws"
+	"github.com/giantswarm/awstpr"
+	microerror "github.com/giantswarm/microkit/error"
+	"github.com/juju/errgo"
+)
+
+type hostedZoneInput struct {
+	Cluster awstpr.CustomObject
+	Client  *route53.Route53
+}
+
+type recordSetInput struct {
+	Cluster      awstpr.CustomObject
+	Client       *route53.Route53
+	Resource     resources.DNSNamedResource
+	Domain       string
+	HostedZoneID string
+}
+
+func (s *Service) createHostedZone(input hostedZoneInput) (*awsresources.HostedZone, error) {
+	hzName, err := hostedZoneName(input.Cluster)
+	if err != nil {
+		return nil, microerror.MaskAny(err)
+	}
+
+	hz := &awsresources.HostedZone{
+		Name:    hzName,
+		Comment: hostedZoneComment(input.Cluster),
+		Client:  input.Client,
+	}
+
+	hzCreated, err := hz.CreateIfNotExists()
+	if err != nil {
+		return nil, microerror.MaskAny(fmt.Errorf("error creating hosted zone '%s'", errgo.Details(err)))
+	}
+
+	if hzCreated {
+		s.logger.Log("debug", fmt.Sprintf("created hosted zone '%s'", hz.Name))
+	} else {
+		s.logger.Log("debug", fmt.Sprintf("hosted zone '%s' already exists, reusing", hz.Name))
+	}
+
+	return hz, nil
+}
+
+func (s *Service) deleteRecordSet(input recordSetInput) error {
+	hzName, err := hostedZoneName(input.Cluster)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	hz, err := awsresources.NewHostedZoneFromExisting(hzName, input.Client)
+	if err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	rs := &awsresources.RecordSet{
+		Client:       input.Client,
+		Resource:     input.Resource,
+		Domain:       input.Domain,
+		HostedZoneID: hz.ID(),
+	}
+
+	if err := rs.Delete(); err != nil {
+		return microerror.MaskAny(err)
+	}
+
+	return nil
+}
+
+func (s *Service) createRecordSet(input recordSetInput) error {
+	// Create DNS records for LB.
+	apiRecordSet := &awsresources.RecordSet{
+		Client:       input.Client,
+		Resource:     input.Resource,
+		Domain:       input.Domain,
+		HostedZoneID: input.HostedZoneID,
+	}
+
+	if err := apiRecordSet.CreateOrFail(); err != nil {
+		return microerror.MaskAny(fmt.Errorf("error registering DNS '%s'", errgo.Details(err)))
+	}
+
+	s.logger.Log("debug", "created or reused DNS record for api")
+
+	return nil
+}
+
+func hostedZoneComment(cluster awstpr.CustomObject) string {
+	return fmt.Sprintf("Hosted zone for cluster %s", cluster.Spec.Cluster.Cluster.ID)
+}
+
+// hostedZoneName removes the first 2 subdomains from the API domain
+// e.g.  apiserver.foobar.aws.giantswarm.io -> aws.giantswarm.io
```

Comment:
Yes, I was undecided between specifying everything in the TPO, and inferring stuff sometimes.

I think we have to strike a balance.